### PR TITLE
[JENKINS-27392] SimpleBuildWrapper

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-27392](https://issues.jenkins-ci.org/browse/JENKINS-27392)

Downstream of https://github.com/jenkinsci/jenkins/pull/1634 and https://github.com/jglick/workflow-plugin/pull/1.

@reviewbybees

### CODE REVIEW

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java [hunk 1]
```diff
@@ -108,7 +110,31 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) {
             }
         }
 
-        return new MaskPasswordsOutputStream(logger, allPasswords);
+        return new FilterImpl(allPasswords);
+    }
+
+    private static final class FilterImpl extends ConsoleLogFilter implements Serializable {
```
Comment (kohsuke) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: Maybe we should promote `@PersistedIn` annotation so that we can use it elsewhere. I find the designation like that useful, as persistence situation affects migration/evolution story of the class.

**Comment by jglick** [comment]
@rseguy @ndeloof anyone get a chance to review this?

**Comment by stephenc** [comment]
:bee:

**Comment by jglick** [comment]
Does this plugin have a maintainer?

**Comment by jglick** [comment]
@reviewbybees done

**Comment by oleg-nenashev** [comment]
IMO #4 should be merged and released before going forward with the core version upgrade

**Comment by jglick** [comment]
Agreed.

**Comment by jglick** [comment]
@oleg-nenashev are you now acting as maintainer?

**Comment by oleg-nenashev** [comment]
Kinda (I'll update the maintainers list later). Let me review it

File: pom.xml [hunk 2]
```diff
@@ -4,7 +4,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.420</version>
+    <version>1.609.1</version>
```
Comment (oleg-nenashev) [inline pom.xml]: I would not say I extremely like it, but I understand the reason

File: pom.xml [hunk 3]
```diff
@@ -43,6 +50,49 @@
...
 
+    <properties>
+        <workflow.version>1.8</workflow.version>
+    </properties>
+    <dependencies>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins.workflow</groupId>
+            <artifactId>workflow-job</artifactId>
+            <version>${workflow.version}</version>
+            <scope>test</scope>
```
Comment (oleg-nenashev) [inline pom.xml]: I think it's better to depend on the aggregator in the test scope

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java [hunk 4]
```diff
@@ -104,7 +102,9 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) {
         if(params != null) {
             for(ParameterValue param : params) {
                 if(config.isMasked(param.getClass().getName())) {
-                    String password = param.createVariableResolver(build).resolve(param.getName());
+                    EnvVars env = new EnvVars();
+                    param.buildEnvironment(build, env);
+                    String password = env.get(param.getName());
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: Seems to be reasonable even if it slightly extends the resolution scope

**Comment by oleg-nenashev** [comment]
I'll send a message to the mailing list. If there's no push-back regarding the core dependency upgrade, I'll merge the change and kick-off the release

File: pom.xml (hunk 3 above)
Comment (jglick) [inline pom.xml]: Also possible, though this carries some complications, for example that you need a `test`-scoped dependency on `git-server` too. I preferred to limit the dependencies to those which are actually being used.

File: src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        });
+    }
+
+    // Copied from credentials-binding-plugin; perhaps belongs in JenkinsRule?
+    private static Set<String> grep(File dir, String text) throws IOException {
+        Set<String> matches = new TreeSet<String>();
+        grep(dir, text, "", matches);
+        return matches;
+    }
+    private static void grep(File dir, String text, String prefix, Set<String> matches) throws IOException {
```
Comment (recena) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: @jglick At least, new line.

File: pom.xml (hunk 3 above)
Comment (recena) [inline pom.xml]: @jglick I thought that the rule would be include `workflow-aggregator`in test scope in order to simplify the dependencies.

File: pom.xml [hunk 6]
```diff
@@ -21,6 +21,13 @@
     <tag>HEAD</tag>
   </scm>
 
+    <licenses>
+        <license>
+            <name>MIT License</name>
+            <url>http://opensource.org/licenses/MIT</url>
+        </license>
+    </licenses>
+
```
Comment (recena) [inline pom.xml]: @jglick Here @oleg-nenashev would say: "Unrelated changes" :smile:

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java [hunk 7]
```diff
@@ -73,13 +77,7 @@ public MaskPasswordsBuildWrapper(List<VarPasswordPair> varPasswordPairs) {
...
     }
 
-    //TODO: Most probably the method is not required after introducing sensitive vars
-    /**
-     * This method is invoked before {@link #makeBuildVariables()} and {@link
-     * #setUp()}.
-     */
-    @Override
-    public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) {
+    @Override public ConsoleLogFilter createLoggerDecorator(Run<?, ?> build) {
```
Comment (KostyaSha) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: Please back override on new line, follow existed style.

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java [hunk 8]
```diff
@@ -112,7 +112,31 @@ public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) {
             }
         }
 
-        return new MaskPasswordsOutputStream(logger, allPasswords);
+        return new FilterImpl(allPasswords);
+    }
+
+    private static final class FilterImpl extends ConsoleLogFilter implements Serializable {
+
+        private static final long serialVersionUID = 1;
```
Comment (KostyaSha) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: `1L` ?

**Comment by KostyaSha** [comment]
Initial coding style has `@Override` on new lines, but in all changed blocks by @jglick they are inlined. Please back to initial style.

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java [hunk 9]
```diff
@@ -140,6 +164,10 @@ public void makeBuildVariables(AbstractBuild build, Map<String, String> variable
         }
     }
 
+    @Override public void setUp(Context context, Run<?, ?> build, FilePath workspace, Launcher launcher, TaskListener listener, EnvVars initialEnvironment) throws IOException, InterruptedException {
```
Comment (lanwen) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: this line too long, please add line breaks

File: src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java (hunk 5 above)
Comment (jglick) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: Helper method used only in the above method, which in a better language would be a local method declaration.

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java (hunk 8 above)
Comment (jglick) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: Does not matter.

File: pom.xml (hunk 3 above)
Comment (jglick) [inline pom.xml]: See above.

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java (hunk 8 above)
Comment (KostyaSha) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: https://google.github.io/styleguide/javaguide.html#s4.8.8-numeric-literals

File: pom.xml (hunk 6 above)
Comment (jglick) [inline pom.xml]: Actually related: I was adding a new file, so wanted to make sure the POM was declaring what license that would be under (which NetBeans takes into account when adding a default license header).

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java (hunk 8 above)
Comment (jglick) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: Yes I know about the syntax for `long` literals, which is not necessary when you are assigning to a `long`-typed field.

File: pom.xml (hunk 6 above)
Comment (KostyaSha) [inline pom.xml]: Makes sense, but => separate commit?

File: src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java (hunk 8 above)
Comment (KostyaSha) [inline src/main/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsBuildWrapper.java]: Also @stephenc contributing.md suggests using exactly `1L` + https://sonar43.spring.io/rules/show/squid:LowerCaseLongSuffixCheck?layout=false
Though, as you wish.

File: pom.xml (hunk 6 above)
Comment (amuniz) [inline pom.xml]: @jglick good IDE that one :smile:

**Comment by amuniz** [comment]
:bee:

**Comment by jglick** [comment]
@reviewbybees done

Leave code formatting comments to the maintainer.

File: pom.xml (hunk 6 above)
Comment (oleg-nenashev) [inline pom.xml]: Yeah, Netbeans is very useful for common Jenkins development flows (special Kudos to Jesse for Stapler plugin & Co)

**Comment by oleg-nenashev** [comment]
@jglick 
I agree with comments regarding annotations. Let's follow the original style. "1L" also makes sense just to follow the common practice. Regarding the line lengths, I'm OK with the current state

File: src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        });
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("p", WorkflowJob.class);
+                WorkflowRun b = p.getLastBuild();
+                assertEquals("TODO cannot keep it out of the closure block, but at least outside users cannot see this; withCredentials does better", new HashSet<String>(Arrays.asList("build.xml", "program.dat")), grep(b.getRootDir(), "s3cr3t"));
+                SemaphoreStep.success("restarting/1", null);
+                story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b));
+                story.j.assertLogContains("printed ******** oops", b);
```
Comment (jtnord) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: you may want to check the log of the node as well as the run. (which at least for timestamps behaves differently)

File: src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+                assertEquals("in build.xml only because it was literally in program text", Collections.singleton("build.xml"), grep(b.getRootDir(), "s3cr3t"));
```
Comment (jtnord) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: woukd it not be better to use a password parameter for the build and enforce it is not here at all?

File: src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java (hunk 10 above)
Comment (jglick) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: Not sure I understand. Slave log? What would that have to do with this?

File: src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java (hunk 11 above)
Comment (jglick) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: Could do that.

File: src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java (hunk 10 above)
Comment (jtnord) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: My bad its an overloaded term and context is king -  its the node of the flow not the Computer Node....

Probably `${JENKINS_URL}/job/p/1/execution/node/2/log/`

Comment (jglick) [inline src/test/java/com/michelin/cio/hudson/plugins/maskpasswords/MaskPasswordsWorkflowTest.java]: Ah, I see. Well this is already covered by the call to `grep` in the next line, I think. (Yes I am not specifically testing the REST API, but that merely streams the `log` file from disk.)

**Comment by oleg-nenashev** [comment]
> I'll send a message to the mailing list. If there's no push-back regarding the core dependency upgrade, I'll merge the change and kick-off the release

Nobody responded. I consider it as "go forward" according to previous discussions with Romain.

File: pom.xml (hunk 3 above)
Comment (oleg-nenashev) [inline pom.xml]: Since there are potential conflicts between Workflow runtime extensions and other stuff, I finally agree that we should minimize the dependency scope to avoid false-positives => no changes needed

**Comment by stephenc** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
:bee: after going through the code

**Comment by oleg-nenashev** [comment]
@reviewbybees done
@jglick Could you squash it?

**Comment by oleg-nenashev** [comment]
Manually integrated as https://github.com/jenkinsci/mask-passwords-plugin/commit/453e8b54b64ad2b54335b1f4265fd90e33d8bf2e
